docs/adr/adr-003-march2022-testnet.md

```diff
@@ -0,0 +1,155 @@
+`StateService` is responsible for fetching state relevant to a user being able to submit a transaction, such as account
+balance, preparing the transaction, and propagating it via `TxSub`. **Bridge** nodes will be responsible for listening
+to `TxSub` and relaying the transactions into the Core mempool.
+
+**@TODO NEEDS MORE RESEARCH**
+
+### Data Availability Sampling during `HeaderSync`
+
+Currently, both **light** and **full* nodes are unable to perform data availability sampling (DAS) while syncing.
+They only begin sampling once the node is synced up to head of chain. 
+
+`HeaderSync` and the `DASer` will be refactored such that the `DASer` will be able to perform sampling on past headers
+as the node is syncing. To do this, the syncing algorithms in both the `DASer` and `HeaderSync` should align so that
+headers received during sync will be propagated to the `DASer` for sampling via an internal pubsub.
+
+The `DASer` will maintain a checkpoint to the last sampled header so that it can continue sampling from the last 
+checkpoint on any new headers.
+
+
+<hr style="border:1px solid gray"> </hr>
+
+## Refactoring
+
+### `HeaderService` becomes main component around which most other services are focused
+Initially, we started with BlockService being the more “important” component during devnet architecture, but overlooked
+some problems with regards to sync (we initially made the decision that a celestia full node would have to be started 
+together at the same time as a core node -- which is the reason for embedding the core node).
+
+This led us to an issue where eventually we needed to connect to an already-running core node and sync from it. We were
+missing a component to do that, so we implemented `HeaderExchange` over the core client (wrapping another interface we 
+had previously created for BlockService called `BlockFetcher`), and we had to do this last minute b/c it wouldn’t work 
+otherwise, leading to a bunch of hacks and other issues (like having to hand the celestia full node a “trusted” hash of
+a header from the already-running chain so that it can sync up to that point and start listening for new headers.
+
+**Proposed new architecture**:
+
+### `BlockService` is only responsible for reconstructing the block from Shares handed to it by the `ShareService`
+Right now, the `BlockService` is in charge of fetching new blocks from the core node, erasure coding them, generating 
+DAH, generating `ExtendedHeader`, broadcasting `ExtendedHeader` to `HeaderSub` network, and storing the block data 
+(after some validation checks).
+
+Instead, we should rely on ShareService sampling to fetch us *enough* shares to reconstruct the block inside of 
+`BlockService`.
+
+### `ShareService` optimizations
+* Implement parallelization for retrieving shares by namespace. This
+[issue](https://github.com/celestiaorg/celestia-node/issues/184) is already being worked on.
+* NMT/Shares/Namespace storage optimizations (**TODO @WONDERTAN**)
```

Comment:
@Wondertan tagging you here so you can fill in the gaps with a sentence or two (please see all `TODO @WONDERTAN` notes in the document)

Reply:
Wouldn't the main storage optimization be that inner nodes won't get stored but recomputed instead? Or is this about other optimizations?

Reply:
> Wouldn't the main storage optimization be that inner nodes won't get stored but recomputed instead?
I still don't understand how this can be implemented at all, but would like to have this implemented. Can we discuss this at some point?

Reply:
>> Wouldn't the main storage optimization be that inner nodes won't get stored but recomputed instead?

> I still don't understand how this can be implemented at all, but would like to have this implemented. Can we discuss this at some point? @liamsi

Sure, we should discuss this at some point. We should do some napkin math to decide if we should prioritize this for launch (incl. incentivized testnet then) or not.
How it would work: you simply recompute the inner nodes on demand and only store the leaves? Why would that be impossible?

Reply:
> you simply recompute the inner nodes on demand and only store the leaves?

huh? How? Try to imagine the request flow starting from data root up to leave and what the logic would look like for a responder that only has leaves.

Reply:
> Try to imagine the request flow starting from data root up to leave and what the logic would look like for a responder that only has leaves.

Here is the naive way:

Case 1) requesting all data for a height for instance:
just return all leaves. requester can recompute everything locally and just compare against root.
Case 2) requesting one leaf (e.g. per index):
Load block of that height (all leaves), recompute the inner nodes of corresponding Merkle tree and respond with leaf + corresponding inclusion proof.
case 3) request all leaves according of one namespace:
like in 2) but return a bunch of leaves and a with the corresponding (recomputed) proof.

A less naive approach would acknowledge the fact that always loading all leaves might be undesirable as well. So what could be a middle ground? Only store _some_ inner nodes, e.g. store the tree in packages of smaller subtrees with their roots (aka some inner nodes) and only recompute the missing inner nodes when necessary. 

I think the latter is how roughly [trillian](https://github.com/google/trillian/tree/master/storage#subtrees) and [diem](https://github.com/diem/diem/tree/main/storage/jellyfish-merkle/src) handle tree storage if I understand correctly. IIRC, diem treats every 4 level binary tree as one inner node and in trillian this is configurable. Both are probably not only there to decrease storage but to optimize for IO.

Reply:
Resolving it as unrelated. Extracted into #271